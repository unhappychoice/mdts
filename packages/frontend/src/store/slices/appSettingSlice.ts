import { createSlice } from '@reduxjs/toolkit';

type DarkMode = 'dark' | 'light' | 'auto';
type ContentMode = 'full' | 'compact';

interface AppSettingState {
  darkMode: DarkMode;
  contentMode: ContentMode;
  fileTreeOpen: boolean;
  outlineOpen: boolean;
}

const initialState: AppSettingState = {
  darkMode: 'auto',
  contentMode: 'compact',
  fileTreeOpen: true,
  outlineOpen: true,
};

const appSettingSlice = createSlice({
  name: 'appSetting',
  initialState,
  reducers: {
    saveAppSetting: (state, action: PayloadAction<{ darkMode: DarkMode, contentMode: ContentMode }>) => {
      state.darkMode = action.payload.darkMode;
      state.contentMode = action.payload.contentMode;
    },
    toggleFileTree: (state) => {
      state.fileTreeOpen = !state.fileTreeOpen;
    },
    toggleOutline: (state) => {
      state.outlineOpen = !state.outlineOpen;
    },
  },
});

export const { saveAppSetting, toggleFileTree, toggleOutline } = appSettingSlice.actions;

export default appSettingSlice.reducer;
