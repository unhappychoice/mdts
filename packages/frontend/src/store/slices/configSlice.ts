import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

export const fetchConfig = createAsyncThunk('config/fetchConfig', async () => {
  return await fetchData<ConfigState>('/api/config', 'json');
});

export const saveConfigToBackend = createAsyncThunk(
  'config/saveConfigToBackend',
  async (config: ConfigState, { dispatch }) => {
    try {
      await fetch('/api/config', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(config),
      });
      dispatch(fetchConfig()); // Re-fetch config after saving
    } catch (error) {
      console.error('Failed to save settings:', error);
      throw error; // Re-throw to allow error handling in component
    }
  }
);

export interface ConfigState {
  fontFamily: string;
  fontFamilyMonospace: string;
  fontSize: number;
  syntaxHighlighterTheme: string;
}

const initialState: ConfigState = {
  fontFamily: 'Roboto',
  fontFamilyMonospace: 'monospace',
  fontSize: 14,
  syntaxHighlighterTheme: 'auto',
};

const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchConfig.fulfilled, (state, action: PayloadAction<ConfigState>) => {
      state.fontFamily = action.payload.fontFamily;
      state.fontFamilyMonospace = action.payload.fontFamilyMonospace;
      state.fontSize = action.payload.fontSize;
      state.syntaxHighlighterTheme = action.payload.syntaxHighlighterTheme;
    });
  },
});

export default configSlice.reducer;
