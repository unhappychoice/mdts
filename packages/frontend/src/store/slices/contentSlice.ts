
import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

interface ContentState {
  content: string;
  loading: boolean;
  error: string | null;
  scrollPosition: number;
}

const initialState: ContentState = {
  content: '',
  loading: true,
  error: null,
  scrollPosition: 0,
};

export const fetchContent = createAsyncThunk(
  'content/fetchContent',
  async (path: string | null) => {
    const url = path ? `/api/markdown/${path}` : '/api/markdown/mdts-welcome-markdown.md';
    const data = await fetchData<string>(url, 'text');
    return data || '';
  }
);

const contentSlice = createSlice({
  name: 'content',
  initialState,
  reducers: {
    setScrollPosition: (state, action: PayloadAction<number>) => {
      state.scrollPosition = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchContent.pending, (state) => {
        if (!state.content) {
          state.loading = true;
        }
        state.error = null;
      })
      .addCase(fetchContent.fulfilled, (state, action) => {
        state.loading = false;
        state.content = action.payload;
      })
      .addCase(fetchContent.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message || 'Failed to fetch content';
      });
  },
});

export const { setScrollPosition } = contentSlice.actions;

export default contentSlice.reducer;
