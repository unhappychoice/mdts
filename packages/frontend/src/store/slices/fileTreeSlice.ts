
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

interface FileTreeItem {
  [key: string]: (FileTreeItem | string)[];
}

interface FileTreeState {
  fileTree: (FileTreeItem | string)[];
  filteredFileTree: (FileTreeItem | string)[];
  searchQuery: string;
  expandedNodes: string[];
  mountedDirectoryPath: string;
  loading: boolean;
  error: string | null;
}

const initialState: FileTreeState = {
  fileTree: [],
  filteredFileTree: [],
  searchQuery: '',
  expandedNodes: [],
  mountedDirectoryPath: '',
  loading: true,
  error: null,
};

const LOCAL_STORAGE_KEY_PREFIX = 'mdts_expanded_nodes_';
const LOCAL_STORAGE_RECENT_PATHS_KEY = 'mdts_recent_paths';
const MAX_RECENT_PATHS = 10;

const saveExpandedNodes = (path: string, nodes: string[]) => {
  try {
    localStorage.setItem(`${LOCAL_STORAGE_KEY_PREFIX}${path}`, JSON.stringify(nodes));
    let recentPaths: string[] = JSON.parse(localStorage.getItem(LOCAL_STORAGE_RECENT_PATHS_KEY) || '[]');
    recentPaths = recentPaths.filter(p => p !== path);
    recentPaths.unshift(path);
    if (recentPaths.length > MAX_RECENT_PATHS) {
      const oldPath = recentPaths.pop();
      if (oldPath) {
        localStorage.removeItem(`${LOCAL_STORAGE_KEY_PREFIX}${oldPath}`);
      }
    }
    localStorage.setItem(LOCAL_STORAGE_RECENT_PATHS_KEY, JSON.stringify(recentPaths));
  } catch (e) {
    console.error('Failed to save expanded nodes to local storage', e);
  }
};

const loadExpandedNodes = (path: string): string[] => {
  try {
    const storedNodes = localStorage.getItem(`${LOCAL_STORAGE_KEY_PREFIX}${path}`);
    return storedNodes ? JSON.parse(storedNodes) : [];
  } catch (e) {
    console.error('Failed to load expanded nodes from local storage', e);
    return [];
  }
};

const filterTree = (tree: (FileTreeItem | string)[], searchQuery: string): (FileTreeItem | string)[] => {
  if (!searchQuery) return tree;

  return tree
    .reduce((acc: (FileTreeItem | string)[], item) => {
      if (typeof item === 'string') {
        const fileName = item.split('/').pop() || '';
        return fileName.toLowerCase().includes(searchQuery.toLowerCase())
          ? [...acc, item]
          : acc;
      } else {
        const key = Object.keys(item)[0];
        const value = item[key];

        const children = Array.isArray(value)
          ? filterTree(value, searchQuery)
          : [];

        return children.length > 0
          ? [...acc, { [key]: children } as FileTreeItem ]
          :acc;
      }
    }, []);
};

export const fetchFileTree = createAsyncThunk(
  'fileTree/fetchFileTree',
  async () => {
    const data = await fetchData<{ fileTree: FileTreeItem[], mountedDirectoryPath: string }>('/api/filetree', 'json');
    return { fileTree: data?.fileTree || [], mountedDirectoryPath: data?.mountedDirectoryPath };
  }
);

const fileTreeSlice = createSlice({
  name: 'fileTree',
  initialState,
  reducers: {
    setSearchQuery: (state, action: { payload: string }) => {
      state.searchQuery = action.payload;
      state.filteredFileTree = filterTree(state.fileTree, action.payload);
    },
    toggleNode: (state, action: { payload: string }) => {
      const path = action.payload;
      if (state.expandedNodes.includes(path)) {
        state.expandedNodes = state.expandedNodes.filter((nodePath) => nodePath !== path);
      } else {
        state.expandedNodes.push(path);
      }
      saveExpandedNodes(state.mountedDirectoryPath, state.expandedNodes);
    },
    setExpandedNodes: (state, action: { payload: string[] }) => {
      state.expandedNodes = action.payload;
      saveExpandedNodes(state.mountedDirectoryPath, state.expandedNodes);
    },
    expandAllNodes: (state, action: { payload: (FileTreeItem | string)[] | null }) => {
      const allItemIds: string[] = [];
      const collectIds = (items: (FileTreeItem | string)[], parentPath: string = '') => {
        items.forEach(item => {
          if (typeof item !== 'string') {
            const key = Object.keys(item)[0];
            const currentPath = parentPath ? `${parentPath}/${key}` : key;
            allItemIds.push(currentPath);
            collectIds(item[key] as (FileTreeItem | string)[], currentPath);
          }
        });
      };
      if (action.payload) {
        collectIds(action.payload);
      }
      state.expandedNodes = allItemIds;
      saveExpandedNodes(state.mountedDirectoryPath, state.expandedNodes);
    },
    setMountedDirectoryPath: (state, action: { payload: string }) => {
      state.mountedDirectoryPath = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchFileTree.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchFileTree.fulfilled, (state, action) => {
        state.loading = false;
        state.fileTree = action.payload.fileTree;
        state.mountedDirectoryPath = action.payload.mountedDirectoryPath;
        state.filteredFileTree = filterTree(action.payload.fileTree, state.searchQuery);
        state.expandedNodes = loadExpandedNodes(action.payload.mountedDirectoryPath);
      })
      .addCase(fetchFileTree.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message || 'Failed to fetch file tree';
      });
  },
});

export const selectFilteredFileTree = (
  fullFileTree: (FileTreeItem | string)[],
  targetPath: string
): (FileTreeItem | string)[] => {
  if (targetPath === '') {
    const result: FileTreeItem[] | string[] = [];
    fullFileTree.forEach(item => {
      if (typeof item === 'string') {
        result.push(item.split('/').pop() || item);
      } else {
        const key = Object.keys(item)[0];
        const newObject: FileTreeItem = {};
        newObject[key.split('/').pop() || key] = item[key];
        result.push(newObject);
      }
    });
    return result;
  }

  const findChildren = (
    currentTree: (FileTreeItem | string)[],
    pathSegments: string[],
    currentSegmentIndex: number
  ): (FileTreeItem | string)[] | null => {
    if (currentSegmentIndex === pathSegments.length) {
      return currentTree;
    }

    const segment = pathSegments[currentSegmentIndex];

    for (const item of currentTree) {
      if (typeof item !== 'string') {
        const key = Object.keys(item)[0];
        const itemSegments = key.split('/');
        if (itemSegments[itemSegments.length - 1] === segment) {
          const children = item[key];
          if (Array.isArray(children)) {
            return findChildren(children, pathSegments, currentSegmentIndex + 1);
          }
        }
      }
    }
    return null;
  };

  const pathSegments = targetPath.split('/').filter(s => s !== '');
  const children = findChildren(fullFileTree, pathSegments, 0);

  if (children) {
    const result: FileTreeItem[] | string[] = [];
    children.forEach(item => {
      if (typeof item === 'string') {
        result.push(item.split('/').pop() || item);
      } else {
        const key = Object.keys(item)[0];
        const newObject: FileTreeItem = {};
        newObject[key.split('/').pop() || key] = item[key];
        result.push(newObject);
      }
    });
    return result;
  }

  return [];
};

export const {
  setSearchQuery,
  toggleNode,
  setExpandedNodes,
  expandAllNodes,
  setMountedDirectoryPath,
} = fileTreeSlice.actions;

export default fileTreeSlice.reducer;
