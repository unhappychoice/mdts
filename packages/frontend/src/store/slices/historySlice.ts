import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AppDispatch } from '../store';

interface HistoryState {
  currentPath: string | null;
  isDirectory: boolean;
}

const initialState: HistoryState = {
  currentPath: null,
  isDirectory: false,
};

const historySlice = createSlice({
  name: 'history',
  initialState,
  reducers: {
    setHistory: (
      state,
      action: PayloadAction<{ path: string | null; isDirectory: boolean }>
    ) => {
      state.currentPath = action.payload.path;
      state.isDirectory = action.payload.isDirectory;
    },
  },
});

export const { setHistory } = historySlice.actions;

export const updateHistoryFromLocation = (pathname: string) => (dispatch: AppDispatch): void => {
  const getPathFromUrl = () => {
    const path = pathname.substring(1);
    if (path === '') return { path: null, isDirectory: false };

    const fileExtensions = ['.md', '.markdown'];
    const isFile = fileExtensions.some((ext) =>
      path.toLowerCase().endsWith(ext)
    );

    return { path: decodeURIComponent(path), isDirectory: !isFile };
  };

  const { path, isDirectory } = getPathFromUrl();
  dispatch(setHistory({ path, isDirectory }));
};

export default historySlice.reducer;
