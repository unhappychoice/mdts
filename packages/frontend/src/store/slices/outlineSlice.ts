import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

export type OutlineItem = {
  id: string;
  content: string;
  level: number;
};

interface OutlineState {
  outline: OutlineItem[];
  loading: boolean;
  error: string | null;
}

const initialState: OutlineState = {
  outline: [],
  loading: true,
  error: null,
};

export const fetchOutline = createAsyncThunk(
  'outline/fetchOutline',
  async (path: string | null) => {
    const url = path
      ? `/api/outline?filePath=${encodeURIComponent(path)}`
      : '/api/outline?filePath=mdts-welcome-markdown.md';
    const data = await fetchData<OutlineItem[]>(url, 'json');
    return data || [];
  }
);

const outlineSlice = createSlice({
  name: 'outline',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchOutline.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchOutline.fulfilled, (state, action) => {
        state.loading = false;
        state.outline = action.payload;
      })
      .addCase(fetchOutline.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message || 'Failed to fetch outline';
      });
  },
});

export default outlineSlice.reducer;
