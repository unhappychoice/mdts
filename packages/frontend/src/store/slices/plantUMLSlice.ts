import { createSlice, createAsyncThunk, createSelector } from '@reduxjs/toolkit';

const hashString = (str: string): string => {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash;
  }
  return Math.abs(hash).toString(36);
};

interface PlantUMLState {
  svgCache: Record<string, string>;
  loading: Record<string, boolean>;
  errors: Record<string, string | null>;
}

export const selectPlantUMLData = createSelector(
  [
    (state: { plantUML: PlantUMLState }) => state.plantUML, 
    (_: { plantUML: PlantUMLState }, diagram: string) => diagram,
  ],
  (plantUMLState, diagram) => {
    const hash = hashString(diagram);
    return {
      hash,
      svg: plantUMLState.svgCache[hash],
      isLoading: plantUMLState.loading[hash],
      error: plantUMLState.errors[hash],
    };
  }
);

const initialState: PlantUMLState = {
  svgCache: {},
  loading: {},
  errors: {},
};

export const generatePlantUMLSvg = createAsyncThunk(
  'plantUML/generateSvg',
  async (diagram: string, { getState, rejectWithValue }) => {
    const hash = hashString(diagram);
    const state = getState() as { plantUML: PlantUMLState };
    
    if (state.plantUML.svgCache[hash]) {
      return { hash, svg: state.plantUML.svgCache[hash] };
    }

    try {
      const isNetlify = window.location.hostname.includes('netlify');
      const endpoint = isNetlify ? '/.netlify/functions/plantuml' : '/api/plantuml/svg';
      
      const response = await fetch(endpoint, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ diagram }),
      });

      if (!response.ok) {
        throw new Error('Failed to render PlantUML diagram');
      }

      const svg = await response.text();
      return { hash, svg };
    } catch (error) {
      return rejectWithValue(error instanceof Error ? error.message : 'Unknown error');
    }
  }
);

const plantUMLSlice = createSlice({
  name: 'plantUML',
  initialState,
  reducers: {
    clearCache: (state) => {
      state.svgCache = {};
      state.loading = {};
      state.errors = {};
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(generatePlantUMLSvg.pending, (state, action) => {
        const hash = hashString(action.meta.arg);
        state.loading[hash] = true;
        state.errors[hash] = null;
      })
      .addCase(generatePlantUMLSvg.fulfilled, (state, action) => {
        const { hash, svg } = action.payload;
        state.svgCache[hash] = svg;
        state.loading[hash] = false;
        state.errors[hash] = null;
      })
      .addCase(generatePlantUMLSvg.rejected, (state, action) => {
        const hash = hashString(action.meta.arg);
        state.loading[hash] = false;
        state.errors[hash] = action.payload as string;
      });
  },
});

export const { clearCache } = plantUMLSlice.actions;
export default plantUMLSlice.reducer;