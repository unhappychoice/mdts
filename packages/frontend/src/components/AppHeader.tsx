import Brightness4Icon from '@mui/icons-material/Brightness4';
import Brightness7Icon from '@mui/icons-material/Brightness7';
import CropFreeIcon from '@mui/icons-material/CropFree';
import FullscreenIcon from '@mui/icons-material/Fullscreen';
import GitHubIcon from '@mui/icons-material/GitHub';
import { AppBar, Box, IconButton, Link, Toolbar, Tooltip } from '@mui/material';
import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { toggleContentMode, toggleDarkMode, } from '../store/slices/appSettingSlice';

interface AppHeaderProps {
  darkMode: boolean;
  contentMode: string;
  handleFileSelect: (path: string) => void;
}

const AppHeader: React.FC<AppHeaderProps> = ({
  darkMode,
  contentMode,
  handleFileSelect,
}) => {
  const dispatch = useDispatch();

  const handleFileSelectClick = useCallback(() => {
    handleFileSelect('');
  }, [dispatch, handleFileSelect]);

  const handleContentModeClick = useCallback(() => {
    dispatch(toggleContentMode());
  }, [dispatch, toggleContentMode]);

  const handleDarkModeClick = useCallback(() => {
    dispatch(toggleDarkMode());
  }, [dispatch, toggleDarkMode]);

  return (
    <AppBar
      position="static"
      elevation={0}
      sx={(theme) => ({
        bgcolor: theme.palette.background.paper,
        color: theme.palette.text.primary,
        borderBottom: `1px solid ${theme.palette.divider}`,
      })}
    >
      <Toolbar>
        <Box sx={{ flexGrow: 1 }}>
          <Tooltip title="Top page">
            <IconButton disableRipple onClick={handleFileSelectClick} color="inherit">
              <img
                src="/logo.svg"
                alt="mdts logo"
                style={{ height: '56px', marginTop: '-8px', marginLeft: '-36px' }}
              />
            </IconButton>
          </Tooltip>
        </Box>
        <Tooltip title={contentMode === 'fixed' ? 'Exit Fullscreen' : 'Fullscreen'}>
          <IconButton onClick={handleContentModeClick} color="inherit" sx={{ mr: 2 }}>
            {contentMode === 'fixed' ? <CropFreeIcon /> : <FullscreenIcon />}
          </IconButton>
        </Tooltip>
        <Tooltip title={darkMode ? 'Switch to Light Mode' : 'Switch to Dark Mode'}>
          <IconButton sx={{ mr: 2 }} onClick={handleDarkModeClick} color="inherit">
            {darkMode ? <Brightness7Icon /> : <Brightness4Icon />}
          </IconButton>
        </Tooltip>
        <Tooltip title="GitHub Repository">
          <IconButton
            color="inherit"
            href="https://github.com/unhappychoice/mdts"
            target="_blank"
            rel="noopener"
            sx={{ mr: 2 }}
          >
            <GitHubIcon />
          </IconButton>
        </Tooltip>
        <Box sx={{ fontSize: '0.9rem', pt: '2px', fontFamily: 'monospace' }}>
          <Tooltip title="Changelog">
            <Link href="https://github.com/unhappychoice/mdts/blob/main/CHANGELOG.md" target="_blank" rel="noopener">
              v{process.env.APP_VERSION}
            </Link>
          </Tooltip>
        </Box>
      </Toolbar>
    </AppBar>
  );
};

export default AppHeader;
