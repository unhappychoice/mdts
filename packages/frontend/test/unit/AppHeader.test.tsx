import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import AppHeader from '../../src/components/AppHeader';
import { toggleContentMode, toggleDarkMode } from '../../src/store/slices/appSettingSlice';

const mockStore = configureStore([]);

jest.mock('../../src/store/slices/appSettingSlice', () => ({
  ...jest.requireActual('../../src/store/slices/appSettingSlice'),
  toggleContentMode: jest.fn(() => ({ type: 'appSetting/toggleContentMode' })),
  toggleDarkMode: jest.fn(() => ({ type: 'appSetting/toggleDarkMode' })),
}));

describe('AppHeader', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      appSetting: {
        darkMode: false,
        contentMode: 'fixed',
      },
    });
    store.dispatch = jest.fn();
  });

  test('renders correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <AppHeader darkMode={false} contentMode="fixed" handleFileSelect={jest.fn()} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('calls handleFileSelect when logo is clicked', () => {
    const handleFileSelectMock = jest.fn();
    render(
      <Provider store={store}>
        <AppHeader darkMode={false} contentMode="fixed" handleFileSelect={handleFileSelectMock} />
      </Provider>
    );
    fireEvent.click(screen.getByAltText('mdts logo'));
    expect(handleFileSelectMock).toHaveBeenCalledWith('');
  });

  test('dispatches toggleContentMode when content mode button is clicked', () => {
    render(
      <Provider store={store}>
        <AppHeader darkMode={false} contentMode="fixed" handleFileSelect={jest.fn()} />
      </Provider>
    );
    fireEvent.click(screen.getByTestId('CropFreeIcon'));
    expect(store.dispatch).toHaveBeenCalledWith(toggleContentMode());
  });

  test('dispatches toggleDarkMode when dark mode button is clicked', () => {
    render(
      <Provider store={store}>
        <AppHeader darkMode={false} contentMode="fixed" handleFileSelect={jest.fn()} />
      </Provider>
    );
    fireEvent.click(screen.getByTestId('Brightness4Icon'));
    expect(store.dispatch).toHaveBeenCalledWith(toggleDarkMode());
  });
});
