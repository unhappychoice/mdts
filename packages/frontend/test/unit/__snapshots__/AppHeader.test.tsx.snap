// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AppHeader renders correctly 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-10054e1-MuiPaper-root-MuiAppBar-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
    >
      <div
        class="MuiBox-root css-1xvvp3o"
      >
        <button
          aria-label="Top page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-1vqxir3-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <img
            alt="mdts logo"
            src="/logo.svg"
            style="height: 56px; margin-top: -8px; margin-left: -36px;"
          />
        </button>
      </div>
      <button
        aria-label="Exit Fullscreen"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-v8k5r1-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="CropFreeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 5v4h2V5h4V3H5c-1.1 0-2 .9-2 2m2 10H3v4c0 1.1.9 2 2 2h4v-2H5zm14 4h-4v2h4c1.1 0 2-.9 2-2v-4h-2zm0-16h-4v2h4v4h2V5c0-1.1-.9-2-2-2"
          />
        </svg>
      </button>
      <button
        aria-label="Switch to Dark Mode"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-v8k5r1-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="Brightness4Icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zM12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6c3.31 0 6 2.69 6 6s-2.69 6-6 6"
          />
        </svg>
      </button>
      <a
        aria-label="GitHub Repository"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-v8k5r1-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        href="https://github.com/unhappychoice/mdts"
        rel="noopener"
        tabindex="0"
        target="_blank"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="GitHubIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 1.27a11 11 0 00-3.48 21.46c.55.09.73-.28.73-.55v-1.84c-3.03.64-3.67-1.46-3.67-1.46-.55-1.29-1.28-1.65-1.28-1.65-.92-.65.1-.65.1-.65 1.1 0 1.73 1.1 1.73 1.1.92 1.65 2.57 1.2 3.21.92a2 2 0 01.64-1.47c-2.47-.27-5.04-1.19-5.04-5.5 0-1.1.46-2.1 1.2-2.84a3.76 3.76 0 010-2.93s.91-.28 3.11 1.1c1.8-.49 3.7-.49 5.5 0 2.1-1.38 3.02-1.1 3.02-1.1a3.76 3.76 0 010 2.93c.83.74 1.2 1.74 1.2 2.94 0 4.21-2.57 5.13-5.04 5.4.45.37.82.92.82 2.02v3.03c0 .27.1.64.73.55A11 11 0 0012 1.27"
          />
        </svg>
      </a>
      <div
        class="MuiBox-root css-gcae57"
      >
        <a
          aria-label="Changelog"
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1orsvah-MuiTypography-root-MuiLink-root"
          data-mui-internal-clone-element="true"
          href="https://github.com/unhappychoice/mdts/blob/main/CHANGELOG.md"
          rel="noopener"
          target="_blank"
        >
          v
        </a>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
